src/parallel/task_scheduler.cpp

```diff
@@ -69,7 +69,12 @@ bool ConcurrentQueue::DequeueFromProducer(ProducerToken &token, shared_ptr<Task>
 
 #else
 struct ConcurrentQueue {
-	std::queue<shared_ptr<Task>> q;
+	// We use a pointer to QueueProducerToken, because
+	// 1. We need to be able to erase the producer token from the queue in the QueueProducerToken destructor
+	// 2. Hashing a pointer is faster than hashing a unique_ptr<foo>
+	// 3. The QueueProducerToken will not move in memory or change (as it is unique per query), so the pointer will
+	// remain the same
+	std::unordered_map<QueueProducerToken *, std::queue<shared_ptr<Task>>> q;
```

Comment:
Actually - can we just move the `queue` into the `QueueProducerToken` instead, given we never need to dequeue a task without a producer token in the implementation without threads? This should simplify the implementation and lifetime management.

Reply:
While I agree that this will result in a cleaner implementation, I feel like this has two disadvantages:
1. Very different mental model compared to the concurrent implementation
2. It is harder to implement something similar to an asynchronous periodic background task using `emscripten_async_call`. This function would periodically pick up a task from the task scheduler regardless of the producer token and execute it for a certain amount of time, similar to how the scheduler's thread pool operates. 
3. I still feel like the scheduler should have a reference to all tasks that will be executed and if I save the `QueueProducerToken` in the `ConcurrentQueue` the lifecycle handling will still be neccessary